File: funnel/paths.py
#!/usr/bin/env python3
"""Single source of funnel paths — topics/<topic>/ layout.

All functions take topic (and engine where needed) and return pathlib.Path.
No legacy fallback: only the new contract.

Central layout:
    topics/<topic>/
      question.txt
      facts.json
      engines/<engine>.json      raw engine output (report + sources)
      atoms/<engine>.json        atoms after atomize.js
      verdicts/<engine>[__rN].json  fact-check verdicts
      selection.json
      pool.json
      pool_raw.json
      pool_resourced.json
      sections.json
      contradictions.json
      audit_rejected.json
      narratives.json
      config.json
      evidence/
      manifest.json
"""
import os
from pathlib import Path

# Repo root holds the bundled `topics/example` only. Real research data lives in
# an external vault selected by RESEARCH_VAULT so the public repo stays clean.
REPO_ROOT = Path(__file__).resolve().parent.parent
ROOT = REPO_ROOT  # back-compat alias for callers that referenced paths.ROOT


# ── data root resolution (RESEARCH_VAULT) ────────────────────────────────────

def _vault_root() -> "Path | None":
    """Return the configured vault root, or None when RESEARCH_VAULT is unset.

    A RESEARCH_VAULT pointing at a missing directory is an error rather than a
    silent fall-through to a repo write.
    """
    v = os.environ.get("RESEARCH_VAULT")
    if not v:
        return None
    p = Path(v).expanduser()
    if not p.exists():
        raise FileNotFoundError(
            f"RESEARCH_VAULT points to a non-existent directory: {p}"
        )
    return p


def topics_root_for(topic: str) -> Path:
    """Resolve the topics/ root holding this topic.

    Precedence:
      1. RESEARCH_VAULT unset       -> repo `topics/` (bundled example / demo / CI).
      2. vault set, topic in vault  -> `$RESEARCH_VAULT/topics`.
      3. vault set, otherwise       -> `$RESEARCH_VAULT/topics` (new topic).

    When a vault is configured it always owns reads and writes, so a run can
    never write into the git-tracked `topics/example` fixture. The bundled
    example is a no-vault demo: run it with RESEARCH_VAULT unset.
    """
    vault = _vault_root()
    if vault is None:
        return REPO_ROOT / "topics"
    return vault / "topics"


# ── topic directory ──────────────────────────────────────────────────────────

def topic_dir(topic: str) -> Path:
    return topics_root_for(topic) / topic


# ── topic input data ─────────────────────────────────────────────────────────

def question(topic: str) -> Path:
    return topic_dir(topic) / "question.txt"


def facts(topic: str) -> Path:
    return topic_dir(topic) / "facts.json"


# ── raw engine output ────────────────────────────────────────────────────────

def engines_dir(topic: str) -> Path:
    return topic_dir(topic) / "engines"


def engine_raw(topic: str, engine: str) -> Path:
    return engines_dir(topic) / f"{engine}.json"


# ── atoms ────────────────────────────────────────────────────────────────────

def atoms_dir(topic: str) -> Path:
    return topic_dir(topic) / "atoms"


def atoms(topic: str, engine: str) -> Path:
    return atoms_dir(topic) / f"{engine}.json"


# ── selection ────────────────────────────────────────────────────────────────

def selection(topic: str) -> Path:
    return topic_dir(topic) / "selection.json"


def select_by_source(topic: str) -> Path:
    return topic_dir(topic) / "select_by_source.json"


# ── verdicts ─────────────────────────────────────────────────────────────────

def verdicts_dir(topic: str) -> Path:
    return topic_dir(topic) / "verdicts"


# ── fact pool ────────────────────────────────────────────────────────────────

def pool(topic: str) -> Path:
    return topic_dir(topic) / "pool.json"


def pool_raw(topic: str) -> Path:
    return topic_dir(topic) / "pool_raw.json"


def pool_resourced(topic: str) -> Path:
    return topic_dir(topic) / "pool_resourced.json"


def sections(topic: str) -> Path:
    return topic_dir(topic) / "sections.json"


def contradictions(topic: str) -> Path:
    return topic_dir(topic) / "contradictions.json"


# ── audit ────────────────────────────────────────────────────────────────────

def audit(topic: str) -> Path:
    return topic_dir(topic) / "audit_rejected.json"


# ── layout (narratives + config) ─────────────────────────────────────────────

def narratives(topic: str) -> Path:
    return topic_dir(topic) / "narratives.json"


def narratives_lock(topic: str) -> Path:
    return topic_dir(topic) / "narratives.lock"


def config(topic: str) -> Path:
    return topic_dir(topic) / "config.json"


def clarity_lock(topic: str) -> Path:
    return topic_dir(topic) / "clarity.lock"


# ── run artifacts ────────────────────────────────────────────────────────────

def evidence_dir(topic: str) -> Path:
    return topic_dir(topic) / "evidence"


def manifest(topic: str) -> Path:
    return topic_dir(topic) / "manifest.json"

File: js/paths.js
"use strict";
// Single source of funnel paths for the JS engine files — mirrors funnel/paths.py.
//
// Layout: topics/<topic>/ holds question.txt, facts.json, engines/<engine>.json,
// atoms/<engine>.json, verdicts/<engine>[__rN].json, pool*.json, narratives.json,
// config.json, evidence/, manifest.json, …
//
// Two roots:
//   • repo/engine root  — where the scripts (prefetch.py, fetch_url.sh, …) live.
//     RESEARCH_STACK_ROOT overrides it; default is the parent of this js/ dir.
//   • data root         — RESEARCH_VAULT. When set, real research topics live
//     there so the public repo keeps only the bundled topics/example.
//
// topics_root_for() precedence matches paths.py exactly:
//   1. RESEARCH_VAULT unset          -> repo topics/ (bundled example)
//   2. vault set, topic in vault      -> $RESEARCH_VAULT/topics
//   3. vault set, topic only in repo  -> repo topics/ (bundled example)
//   4. vault set, topic in neither    -> $RESEARCH_VAULT/topics (new topic)
const path = require("path");
const fs = require("fs");
const os = require("os");

function expand(p) {
  if (!p) return p;
  if (p === "~") return os.homedir();
  if (p.startsWith("~/")) return path.join(os.homedir(), p.slice(2));
  return p;
}

// Repo/engine root holds the bundled `topics/example` only.
const REPO_ROOT = process.env.RESEARCH_STACK_ROOT
  ? path.resolve(expand(process.env.RESEARCH_STACK_ROOT))
  : path.resolve(__dirname, "..");
const ROOT = REPO_ROOT; // back-compat alias for callers that referenced ROOT

// ── data root resolution (RESEARCH_VAULT) ───────────────────────────────────
function vaultRoot() {
  // Return the configured vault root, or null when RESEARCH_VAULT is unset.
  // A RESEARCH_VAULT pointing at a missing directory is an error rather than a
  // silent fall-through to a repo write.
  const v = process.env.RESEARCH_VAULT;
  if (!v) return null;
  const p = path.resolve(expand(v));
  if (!fs.existsSync(p)) {
    throw new Error(`RESEARCH_VAULT points to a non-existent directory: ${p}`);
  }
  return p;
}

function topicsRootFor(topic) {
  const vault = vaultRoot();
  if (vault === null) return path.join(REPO_ROOT, "topics");
  if (fs.existsSync(path.join(vault, "topics", topic))) {
    return path.join(vault, "topics");
  }
  if (fs.existsSync(path.join(REPO_ROOT, "topics", topic))) {
    return path.join(REPO_ROOT, "topics");
  }
  return path.join(vault, "topics");
}

// ── topic directory ─────────────────────────────────────────────────────────
function topicDir(topic) {
  return path.join(topicsRootFor(topic), topic);
}

// ── topic input data ────────────────────────────────────────────────────────
function question(topic) {
  return path.join(topicDir(topic), "question.txt");
}
function facts(topic) {
  return path.join(topicDir(topic), "facts.json");
}

// ── raw engine output ───────────────────────────────────────────────────────
function enginesDir(topic) {
  return path.join(topicDir(topic), "engines");
}
function engineRaw(topic, engine) {
  return path.join(enginesDir(topic), `${engine}.json`);
}

// ── atoms ───────────────────────────────────────────────────────────────────
function atomsDir(topic) {
  return path.join(topicDir(topic), "atoms");
}
function atoms(topic, engine) {
  return path.join(atomsDir(topic), `${engine}.json`);
}

// ── selection ───────────────────────────────────────────────────────────────
function selection(topic) {
  return path.join(topicDir(topic), "selection.json");
}
function selectBySource(topic) {
  return path.join(topicDir(topic), "select_by_source.json");
}

// ── verdicts ────────────────────────────────────────────────────────────────
function verdictsDir(topic) {
  return path.join(topicDir(topic), "verdicts");
}
function verdict(topic, engine, round) {
  const suffix = round && Number(round) > 1 ? `__r${round}` : "";
  return path.join(verdictsDir(topic), `${engine}${suffix}.json`);
}

// ── fact pool ───────────────────────────────────────────────────────────────
function pool(topic) {
  return path.join(topicDir(topic), "pool.json");
}
function poolRaw(topic) {
  return path.join(topicDir(topic), "pool_raw.json");
}
function poolResourced(topic) {
  return path.join(topicDir(topic), "pool_resourced.json");
}
function sections(topic) {
  return path.join(topicDir(topic), "sections.json");
}
function contradictions(topic) {
  return path.join(topicDir(topic), "contradictions.json");
}

// ── audit ───────────────────────────────────────────────────────────────────
function audit(topic) {
  return path.join(topicDir(topic), "audit_rejected.json");
}

// ── layout (narratives + config) ────────────────────────────────────────────
function narratives(topic) {
  return path.join(topicDir(topic), "narratives.json");
}
function narrativesLock(topic) {
  return path.join(topicDir(topic), "narratives.lock");
}
function config(topic) {
  return path.join(topicDir(topic), "config.json");
}
function clarityLock(topic) {
  return path.join(topicDir(topic), "clarity.lock");
}
function clarityFindings(topic) {
  return path.join(topicDir(topic), "clarity_findings.json");
}

// ── run artifacts ───────────────────────────────────────────────────────────
function evidenceDir(topic) {
  return path.join(topicDir(topic), "evidence");
}
function manifest(topic) {
  return path.join(topicDir(topic), "manifest.json");
}

module.exports = {
  ROOT,
  REPO_ROOT,
  topicsRootFor,
  topicDir,
  question,
  facts,
  enginesDir,
  engines: enginesDir, // alias matching the paths.py accessor shorthand
  engineRaw,
  atomsDir,
  atoms,
  selection,
  selectBySource,
  verdictsDir,
  verdict,
  pool,
  poolRaw,
  poolResourced,
  sections,
  contradictions,
  audit,
  narratives,
  narrativesLock,
  config,
  clarityLock,
  clarityFindings,
  evidenceDir,
  manifest,
};
